sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1208,9 +1208,38 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
             reduceLeftOption(And).map(Filter(_, left)).getOrElse(left)
           val newRight = rightJoinConditions.
             reduceLeftOption(And).map(Filter(_, right)).getOrElse(right)
-          val newJoinCond = commonJoinCondition.reduceLeftOption(And)
-
-          Join(newLeft, newRight, joinType, newJoinCond)
+          val (newJoinConditions, others) =
+            commonJoinCondition.partition(canEvaluateWithinJoin)
+          val newJoinCond = newJoinConditions.reduceLeftOption(And)
+          val newJoinType =
+            if (commonJoinCondition.nonEmpty && newJoinCond.isEmpty) {
+              if (SQLConf.get.crossJoinEnabled) {
+                // if condition expression is unevaluable, it will be removed from
+                // the new join conditions, if all conditions is unevaluable, we should
+                // change the join type to CrossJoin.
+                logWarning(s"The whole commonJoinCondition:$commonJoinCondition of the join " +
+                  "plan is unevaluable, it will be ignored and the join plan will be " +
+                  s"turned to cross join. This plan shows below:\n $j")
+                Cross
+              } else {
+                // if the crossJoinEnabled is false, an AnalysisException will throw by
+                // [[CheckCartesianProducts]], we throw firstly here for better readable
+                // information.
+                throw new AnalysisException("Detected the whole commonJoinCondition:" +
+                  s"$commonJoinCondition of the join plan is unevaluable, we need to cast the " +
+                  "join to cross join by setting the configuration variable " +
+                  s"${SQLConf.CROSS_JOINS_ENABLED.key}=true")
+              }
+            } else {
+              joinType
+            }
+
+          val join = Join(newLeft, newRight, newJoinType, newJoinCond)
+          if (others.nonEmpty && joinType.isInstanceOf[InnerLike]) {
+            Filter(others.reduceLeft(And), join)
+          } else {
+            join
```

Comment:
this means that we are removing without doing anything the condition when we have a SemiJoin. This is wrong. All this logic can be applied only to the Inner case. In the other cases, this fix is wrong. Moreover, please add a UT to enforce the correctness in the case LeftSemi join, so we can be sure that a wrong fix doesn't go in. Thanks.

Reply:
Thanks, I'll do more test on the SemiJoin here, but as currently test over PySpark, this is not wrong, maybe I misunderstand you two `wrong` means, is your `wrong` means correctness or just benchmark regression?
![image](https://user-images.githubusercontent.com/4833765/45043269-4ba20c80-b09f-11e8-84dc-a1f3ff416303.png)

Reply:
It means that in the left_semi join the output of the Join operator should contain only the attributes from the left side, so attributes from the right side should not be referenced after the join. Therefore the plan should be invalid. I am a bit surprised that works, it would be great to understand why. Thanks.

Reply:
Thanks @mgaido91 and @dilipbiswal !
I fix this in 63fbcce. The mainly problem is semi join in both deterministic and non-deterministic condition, filter after semi join will fail. Also add more tests both on python and scala side, including semi join, inner join and complex scenario described below.
It makes the strategy difficult to read after considering left semi, so in 63fbcce I split the logic of semi join and inner join.

Reply:
```
 I am a bit surprised that works, it would be great to understand why. Thanks.
```
Sorry for the bad test, that's too special and the result just right by accident. The original implement will make all semi join return `[]` in PySpark.